### DISCUSSION TITLE
Creating Local Package Fails

### DISCUSSION THREAD

**Discussion by brianhyder** [aws-powertools/powertools-lambda-typescript#1519] [Q&A] [answered]
Hey,

  I pulled the latest from main this morning to get the latest changes for idempotency.  After doing so, I ran `npm run package` as this is what I had been doing to create the local packages for testing idempotency ahead of its release.  I am now getting failures.  I saw that #1504 recently went in and was wondering if the process for creating the package file (.tgz) has now changed.  Thank you in advance for the help!

Output:
```
aws-lambda-powertools-typescript % npm run package

> aws-lambda-powertools-typescript@0.0.1 package
> npm run package -ws

npm ERR! Lifecycle script `package` failed with error: 
npm ERR! Error: Missing script: "package"

To see a list of scripts, run:
  npm run 
npm ERR!   in workspace: @aws-lambda-powertools/commons@1.9.0 
npm ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/packages/commons 
npm ERR! Lifecycle script `package` failed with error: 
npm ERR! Error: Missing script: "package"

To see a list of scripts, run:
  npm run 
npm ERR!   in workspace: @aws-lambda-powertools/logger@1.9.0 
npm ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/packages/logger 
npm ERR! Lifecycle script `package` failed with error: 
npm ERR! Error: Missing script: "package"

To see a list of scripts, run:
  npm run 
npm ERR!   in workspace: @aws-lambda-powertools/metrics@1.9.0 
npm ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/packages/metrics 
npm ERR! Lifecycle script `package` failed with error: 
npm ERR! Error: Missing script: "package"

To see a list of scripts, run:
  npm run 
npm ERR!   in workspace: @aws-lambda-powertools/tracer@1.9.0 
npm ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/packages/tracer 
npm ERR! Lifecycle script `package` failed with error: 
npm ERR! Error: Missing script: "package"

To see a list of scripts, run:
  npm run 
npm ERR!   in workspace: @aws-lambda-powertools/parameters@1.9.0 
npm ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/packages/parameters 
npm ERR! Lifecycle script `package` failed with error: 
npm ERR! Error: Missing script: "package"

To see a list of scripts, run:
  npm run 
npm ERR!   in workspace: @aws-lambda-powertools/idempotency@1.9.0 
npm ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/packages/idempotency 
npm ERR! Lifecycle script `package` failed with error: 
npm ERR! Error: Missing script: "package"

To see a list of scripts, run:
  npm run 
npm ERR!   in workspace: docs@1.9.0 
npm ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/docs/snippets 

> layers@1.9.0 package
> echo 'Not applicable'

Not applicable⠂⠂⠂⠂⠂) ⠸ : ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/docs/snippets
npm ERR! Lifecycle script `package` failed with error: 
npm ERR! Error: Missing script: "package"

To see a list of scripts, run:
  npm run 
npm ERR!   in workspace: cdk-sample@1.9.0 
npm ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/examples/cdk 
npm ERR! Lifecycle script `package` failed with error: 
npm ERR! Error: Missing script: "package"

To see a list of scripts, run:
  npm run 
npm ERR!   in workspace: sam-example@1.9.0 
npm ERR!   at location: /Users/someuser/code/aws-lambda-powertools-typescript/examples/sam
```

**Comment by dreamorosi** [comment, marked as answer]
Hi @brianhyder, you are correct, we tweaked the install, build, and package workflows for the repo [in a recent PR](https://github.com/aws-powertools/powertools-lambda-typescript/pull/1504).

When starting from a fresh clone, you can run `npm run setup-local`, this command will run:
- `npm ci` to install the dependencies for the whole workspace
- `npm run build` which will compile the files from TS to JS
- `npm run init-environment` which installs the commit hooks (not needed for the purposes of this explanation, but adding it for the sake of being complete)

If instead you are starting from an already cloned repo, I'd still recommend you to run the first two commands above (`npm ci` followed by `npm run build`) or the same `npm run setup-local`. This will ensure that you have the latest dependencies and latest build configs/artifacts.

At this point you can pack/bundle each one of the utilities individually with `npm pack -w packages/idempotency` or all of them together with `npm pack -ws` (note that running with `-ws` will also output other packages that are part of the workspace but are not usually published to npm).

My recommendation to avoid interoperability issues, would be to build only one package and use the other ones available from npm.

After building you should run `git restore .`, this reverts any changes to the `package.json` files that is made during the packaging.

Normally all these commands are done in our CI [here](https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/workflows/make-release.yml#L35-L40), and we optimized for this type of workflow.

Let me know if this helps!

**Reply by dreamorosi** [reply]
I should probably document this in the `CONTRIBUTING` guidelines.

**Reply by brianhyder** [reply]
Ah, thank you for the clarity.  I'm going to try it in the next hour.  That would definitely help.  I looked in the README but didn't see anything.

**Reply by brianhyder** [reply]
The root _[package.json](https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/package.json#L22)_ file still has the `package` script defined.  Did you all mean to remove it or is it still being used for something?

**Reply by dreamorosi** [reply]
Yes, fair point. It's not a workflow that we expected our customers to do up until now but I can see how it'd be useful for maintainers and collaborators. I've opened an issue (#1525) and will be opening a PR soon. 

(I'll be removing that script, good spot)
